### PR TITLE
feat(express_next): wrap next middleware in express

### PR DESCRIPTION
**Pull request by haimrait**
part of https://github.com/epsagon/epsagon-node/pull/350

Screen Shot 2020-08-31 at 20 19 51

### CODE REVIEW

File: src/trace_context.js [hunk 1]
```diff
@@ -68,11 +68,25 @@ function RunInContext(createTracer, handle) {
...
+    return null;
+}
+
+/**
+ * Creates a reference to epsagonIdentifier
+ * @param {Number} epsagonIdentifier sets the reference to this epsagonIdentifier
+ */
+function setTraceToEpsagonId(epsagonIdentifier) {
+    tracers[epsagonIdentifier] = get(epsagonIdentifier);
 }
```
Comment (galbash) [inline src/trace_context.js:90]: why get with `epsagonIdentifier`? don't you want to use the `executionAsyncId()` here?

File: src/wrappers/express.js [hunk 2]
```diff
@@ -30,7 +31,11 @@ function expressMiddleware(req, res, next) {
     tracer.restart();
     let expressEvent;
     const startTime = Date.now();
+    const epsagonIdentifier = uuid4();
     try {
+        // Add epsagon id to
+        req.epsagonId = epsagonIdentifier;
+        tracer.getTrace = () => traceContext.get(epsagonIdentifier);
```
Comment (galbash) [inline src/wrappers/express.js:38]: this will be problematic in a multi request environment - tracer is a singleton!

Comment (galbash) [reply src/wrappers/express.js:38]: everything in this case must be managed on the request instead of in the singleton.

File: src/trace_context.js [hunk 3]
```diff
@@ -68,11 +68,25 @@ function RunInContext(createTracer, handle) {
...
+    }
+    return null;
+}
+
+/**
+ * Creates a reference to epsagonIdentifier
+ * @param {Number} epsagonIdentifier sets the reference to this epsagonIdentifier
+ */
+function setTraceToEpsagonId(epsagonIdentifier) {
+    tracers[epsagonIdentifier] = get(epsagonIdentifier);
```
Comment (ranrib) [inline src/trace_context.js:89]: It seems like this should never work because when you set it, `get(epsagonIdentifier)` should be null

File: src/wrappers/express.js (hunk 2 above)
Comment (ranrib) [reply src/wrappers/express.js:38]: This line should be instead of line 111

File: src/wrappers/express.js [hunk 4]
```diff
@@ -75,6 +80,25 @@ function expressMiddleware(req, res, next) {
...
+function nextWrapper(req, next) {
+    const originalNext = next;
+    return function internalNextWrapper(error) {
+        if (error) {
+            utils.debugLog('Epsagon Next - middleware executed');
+            utils.debugLog(error);
+        }
+        traceContext.setTraceToEpsagonId(req.epsagonId);
+        originalNext(...arguments);
+        traceContext.setTraceToEpsagonId(req.epsagonId);
```
Comment (ranrib) [inline src/wrappers/express.js:98]: Why is this line needed before and after?

Comment (haimrait) [reply src/wrappers/express.js:97]: The first line set trace before the middleware is running
The second line updates trace if needed

File: src/trace_context.js (hunk 1 above)
Comment (haimrait) [reply src/trace_context.js:90]: get(epsagonIdentifier) calls
`function get(epsagonIdentifier) {
    if (tracers[asyncHooks.executionAsyncId()]) {
        return tracers[asyncHooks.executionAsyncId()];
    } if (epsagonIdentifier) {
        return tracers[epsagonIdentifier];
    }
    return null;
}`

File: src/trace_context.js [hunk 5]
```diff
@@ -68,11 +68,25 @@ function RunInContext(createTracer, handle) {
...
+        return tracers[epsagonIdentifier];
+    }
+    return null;
+}
+
+/**
+ * Creates a reference to epsagonIdentifier
+ * @param {Number} epsagonIdentifier sets the reference to this epsagonIdentifier
+ */
+function setTraceToEpsagonId(epsagonIdentifier) {
```
Comment (galbash) [inline src/trace_context.js:88]: I think you don't need this now (any change to this file actually)

File: src/wrappers/express.js [hunk 6]
```diff
@@ -90,10 +137,12 @@ function expressWrapper(wrappedFunction) {
...
-            (req, res, next) => traceContext.RunInContext(
-                tracer.createTracer,
-                () => expressMiddleware(req, res, next)
-            )
+            (req, res, next) => {
+                traceContext.RunInContext(
+                    tracer.createTracer,
+                    () => expressMiddleware(req, res, next)
+                );
+            }
```
Comment (galbash) [inline src/wrappers/express.js:145]: this breaks the return value

File: src/wrappers/express.js [hunk 7]
```diff
@@ -75,6 +76,52 @@ function expressMiddleware(req, res, next) {
...
+ * Wraps express use function
+ * @param {*} original - original use function
+ * @returns {function} - wrapped use function
+ */
+function useWrapper(original) {
+    return function internalUseWrapper() {
+        // Check if we have middleware
+        if (arguments[1]) {
+            arguments[1] = middlewareWrapper(arguments[1]);
+        }
```
Comment (galbash) [inline src/wrappers/express.js:120]: I think the possible parameters to use are a bit more complex:
http://expressjs.com/en/4x/api.html#router.use

**Review by ranrib — APPROVED** [review]

**Review by galbash — APPROVED** [review]
